### ISSUE TITLE
Zammad image service does not replace avatars (LDAP)

### ISSUE THREAD

**Issue by Schluggi** [zammad/zammad#3358]
### Infos:

* Used Zammad version: 3.6
* Installation method (source, package, ..): [package](https://docs.zammad.org/en/latest/install/ubuntu.html)
* Operating system: Ubuntu 20.04.1 LTS
* Database + version: PostgreSQL 12.5
* Elasticsearch version: 7.10.1
* Browser + version: Firefox 84.0

### Expected behavior:

* The image from the Zammad image service (images.zammad.com) should be used as user avatar

### Actual behavior:

* The default user avatar is displayed

### Steps to reproduce the behavior:

* Setup LDAP integration
* Sync users including there mail addresses 

Yes I'm sure this is a bug and no feature request or a general question.

### What i found out
- When I request the image api directly via my python script, I get the correct image back.
  ```python3
  from requests import post
  import shutil
  
  resp = post('https://images.zammad.com/api/v1/person/image', data={'email': '<user mail address>'}, stream=True)
  
  with open('/tmp/out.jpeg', 'wb') as of:
	  shutil.copyfileobj(resp.raw, of)
  ```

- I can't find a log entry from the image service that it is even trying to get an image.

**Comment by MrGeneration** [comment]
I'm not sure if this is really a bug or rather an enhancement.
Zammad doesn't seem to try that at all, I honestly can't remember if that *ever* worked actually.

I would see this as an enhancement, what do you think Thorsten?

**Comment by thorsteneckel** [comment]
Hi @Schluggi and @MrGeneration - you are right. There are no Avatars fetched for users when they are synced from LDAP. In fact this is true for all imports (e.g. OTRS/Zendesk/...). This works as designed. Let me elaborate: User imports from external sources have the potential to create a lot of Users in Zammad. At the same time it's almost always desirable for an import to finish as fast as possible. On the other hand it's slow and error prone to look up and store an attachment for each record. Therefore we decided to disable the Avatar creation in the context of bulk actions such as imports.

Here's how: Whenever a import saves a record it enables the BulkImportInfo:
https://github.com/zammad/zammad/blob/f6f1b37250b00f48e8fccb5db65b039f8a992fa7/lib/sequencer/unit/import/common/model/save.rb#L25-L33

Whenever a User is [created](https://github.com/zammad/zammad/blob/f6f1b37250b00f48e8fccb5db65b039f8a992fa7/app/models/user.rb#L46) or [updated](https://github.com/zammad/zammad/blob/f6f1b37250b00f48e8fccb5db65b039f8a992fa7/app/models/user.rb#L48) usually a callback to run the method `avatar_for_email_check` is executed but only if the `BulkImportInfo` is not enabled. Therefore the avatar sync is skipped.

However, you can start a manual Avatar sync for all Users synced from LDAP by running the following command via the 
[Zammad rails console](https://docs.zammad.org/en/latest/admin-console.html#start-zammad-s-rails-console):

```ruby
ExternalSync.where(source: 'Ldap::User').pluck(:o_id).each do |user_id|
  user = User.lookup(id: user_id)
  next if user.blank?

  avatar = Avatar.auto_detection(
    object:        'User',
    o_id:          user.id,
    url:           user.email,
    source:        'app',
    updated_by_id: user.updated_by_id,
    created_by_id: user.updated_by_id,
  )

  next if !avatar

  user.update!(image: avatar.store_hash)
end
```

As this works as designed I'll close this issue.

**Comment by Schluggi** [comment]
> However, you can start a manual Avatar sync for all Users synced from LDAP by running the following command via the
> [Zammad rails console](https://docs.zammad.org/en/latest/admin-console.html#start-zammad-s-rails-console):
> 
> ```ruby
> ExternalSync.where(object: 'Ldap::User').pluck(:o_id).each do |user_id|
>   user = User.lookup(id: user_id)
>   next if user.blank?
> 
>   avatar = Avatar.auto_detection(
>     object:        'User',
>     o_id:          user.id,
>     url:           user.email,
>     source:        'app',
>     updated_by_id: user.updated_by_id,
>     created_by_id: user.updated_by_id,
>   )
> 
>   next if !avatar
> 
>   user.update!(image: avatar.store_hash)
> end
> ```
> 
> As this works as designed I'll close this issue.

Thanks for this script, but it doesn't seem to work. After i pasted it, I get an empty list back and nothing else happens.
The avatars are still the same as before.

![image](https://user-images.githubusercontent.com/43608073/104293053-dc822180-54bd-11eb-8b93-d59ef2d16840.png)

**Comment by MrGeneration** [comment]
Hello,

please replace `object: 'Ldap::User'` by `source: 'Ldap::User'` and you should be good to go.
If it still doesn't work, please create a post on the Zammad community as this is a technical question.

Thank you.

**Comment by Schluggi** [comment]
It works! Thanks :tada:

**Comment by picazi** [comment]
Hi,

Sorry for reopening this issue. I tried the code, and I do get an error that the image from Active-Directory could not be found. I don't know where the problem is. I am using: Zammad 5.3.1-1677147662.62001c62.bullseye
 
Import:
irb(main):051:1* ExternalSync.where(source: 'Ldap::User').pluck(:o_id).each do |user_id|
irb(main):052:1*   user = User.lookup(id: user_id)
irb(main):053:1*   next if user.blank?
irb(main):054:1*
irb(main):055:2*   avatar = Avatar.auto_detection(
irb(main):056:2*     object:        'User',
irb(main):057:2*     o_id:          user.id,
irb(main):058:2*     url:           user.email,
irb(main):059:2*     source:        'app',
irb(main):060:2*     updated_by_id: user.updated_by_id,
irb(main):061:2*     created_by_id: user.updated_by_id,
irb(main):062:1*   )
irb(main):063:1*
irb(main):064:1*   next if !avatar
irb(main):065:1*
irb(main):066:1*   user.update!(image: avatar.store_hash)
irb(main):067:0> end
Can't fetch image for 'xxx@xxx.xxx.de' (maybe no avatar available), http code: 404
Can't fetch image for 'xxx@xxx.xxx.de' (maybe no avatar available), http code: 404
Can't fetch image for 'xxx@xxx.xxx.de' (maybe no avatar available), http code: 404
Can't fetch image for 'xxx@xxx.xxx.de' (maybe no avatar available), http code: 404
...

Successful import for one user (production.log):
`[2023-03-30T13:43:55.484325 #482-14114180]  INFO -- : Started GET "/api/v1/users/search?sort_by=created_at&query=matt&limit=50&full=true&_=1680165050753" for xxx.xxx.xxx.xxx at 2023-03-30 13:43:55 +0200
I, [2023-03-30T13:43:55.488794 #482-14114180]  INFO -- : Processing by UsersController#search as JSON
I, [2023-03-30T13:43:55.488867 #482-14114180]  INFO -- :   Parameters: {"sort_by"=>"created_at", "query"=>"matt", "limit"=>"50", "full"=>"true", "_"=>"1680165050753"}`

I checked the image (size, width, and height) and could not find a difference from the other user images. But I discovered that it is not the image from the Attribute thumbnailPhoto. I don't know where this image comes from... But however, I would like to sync the attribute thumbnailPhoto from Active-Directory with Zammd Avatar images.

It would be great to get some feedback.

And I would like a say big thanks for all your work! Zammad is fantastic and works incredibly great for us!!
